### PR TITLE
Fix TAP validation mode

### LINKED ISSUE #15902: panic: not implemented from --mode validate
*Title*: Admin API TAP sockets fail on validate

*Description*:
When validating a static config file using `--mode validate` with tap transport sockets wrapping any other transport socket (tried with both raw buffer and tls) then envoy crashes with a `panic: not implemented` from https://github.com/envoyproxy/envoy/blob/main/source/server/config_validation/admin.cc#L13

*Repro steps*:
run envoy in validate mode with an admin api tap transport socket

```
[2021-04-09 14:35:15.911][35383][critical][assert] [source/server/config_validation/admin.cc:13] panic: not implemented
  [2021-04-09 14:35:15.911][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:104] Caught Aborted, suspect faulting address 0x8a37
  [2021-04-09 14:35:15.911][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:91] Backtrace (use tools/stack_decode.py to get line numbers):
  [2021-04-09 14:35:15.911][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:92] Envoy version: d6a4496e712d7a2335b26e2f76210d5904002c26/1.17.1/Clean/RELEASE/BoringSSL
  [2021-04-09 14:35:15.912][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #0: __restore_rt [0x7f0ad4086980]
  [2021-04-09 14:35:15.918][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #1: Envoy::Extensions::Common::Tap::AdminHandler::AdminHandler() [0x56291892ab7d]
  [2021-04-09 14:35:15.925][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #2: std::__1::__function::__func<>::operator()() [0x56291892c850]
  [2021-04-09 14:35:15.932][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #3: Envoy::Singleton::ManagerImpl::get() [0x56291968bc2c]
  [2021-04-09 14:35:15.939][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #4: Envoy::Singleton::Manager::getTyped<>() [0x56291892a97e]
  [2021-04-09 14:35:15.946][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #5: Envoy::Extensions::Common::Tap::AdminHandler::getSingleton() [0x56291892a867]
  [2021-04-09 14:35:15.953][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #6: Envoy::Extensions::Common::Tap::ExtensionConfigBase::ExtensionConfigBase() [0x562918928df3]
  [2021-04-09 14:35:15.961][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #7: Envoy::Extensions::TransportSockets::Tap::TapSocketFactory::TapSocketFactory() [0x562918928688]
  [2021-04-09 14:35:15.967][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #8: Envoy::Extensions::TransportSockets::Tap::UpstreamTapSocketConfigFactory::createTransportSocketFactory() [0x5629189228cf]
  [2021-04-09 14:35:15.974][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #9: Envoy::Upstream::createTransportSocketFactory() [0x56291984865d]
  [2021-04-09 14:35:15.981][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #10: Envoy::Upstream::ClusterImplBase::ClusterImplBase() [0x562919848b9b]
  [2021-04-09 14:35:15.988][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #11: Envoy::Upstream::StrictDnsClusterImpl::StrictDnsClusterImpl() [0x56291986c86b]
  [2021-04-09 14:35:15.995][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #12: std::__1::__shared_ptr_emplace<>::__shared_ptr_emplace<>() [0x56291986f556]
  [2021-04-09 14:35:16.003][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #13: Envoy::Upstream::StrictDnsClusterFactory::createClusterImpl() [0x56291986da22]
  [2021-04-09 14:35:16.011][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #14: Envoy::Upstream::ClusterFactoryImplBase::create() [0x56291986b1ca]
  [2021-04-09 14:35:16.018][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #15: Envoy::Upstream::ClusterFactoryImplBase::create() [0x56291986a5ee]
  [2021-04-09 14:35:16.025][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #16: Envoy::Upstream::ProdClusterManagerFactory::clusterFromProto() [0x5629196b158d]
  [2021-04-09 14:35:16.032][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #17: Envoy::Upstream::ClusterManagerImpl::loadCluster() [0x5629196a7492]
  [2021-04-09 14:35:16.039][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #18: Envoy::Upstream::ClusterManagerImpl::ClusterManagerImpl() [0x5629196a4a55]
  [2021-04-09 14:35:16.047][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #19: Envoy::Upstream::ValidationClusterManagerFactory::clusterManagerFromProto() [0x56291963510d]
  [2021-04-09 14:35:16.054][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #20: Envoy::Server::Configuration::MainImpl::initialize() [0x56291992736f]
  [2021-04-09 14:35:16.061][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #21: Envoy::Server::ValidationInstance::initialize() [0x56291962f3ab]
  [2021-04-09 14:35:16.068][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #22: Envoy::Server::ValidationInstance::ValidationInstance() [0x56291962e25f]
  [2021-04-09 14:35:16.075][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #23: Envoy::Server::validateConfig() [0x56291962d695]
  [2021-04-09 14:35:16.084][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #24: Envoy::MainCommonBase::run() [0x562917e5be5d]
  [2021-04-09 14:35:16.091][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #25: Envoy::MainCommon::main() [0x562917e5c627]
  [2021-04-09 14:35:16.097][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #26: main [0x562917e5a9dc]
  [2021-04-09 14:35:16.098][35383][critical][backtrace] [bazel-out/k8-opt/bin/source/server/_virtual_includes/backtrace_lib/server/backtrace.h:96] #27: __libc_start_main [0x7f0ad3ca4bf7]
```

```yaml
  - connect_timeout: 5s
    lb_policy: round_robin
    load_assignment:
      cluster_name: service
      endpoints:
      - lb_endpoints:
          endpoint:
            address:
              socket_address:
                address: "service"
                port_value: 443
    name: api-gateway
    transport_socket:
      name: envoy.transport_sockets.tap
      typed_config:
        '@type': type.googleapis.com/envoy.extensions.transport_sockets.tap.v3.Tap
        common_config:
          admin_config:
            config_id: api-gateway
        transport_socket:
          name: envoy.transport_sockets.tls
          typed_config:
            '@type': type.googleapis.com/envoy.extensions.transport_sockets.tls.v3.UpstreamTlsContext
            allow_renegotiation: true
            common_tls_context:
              tls_certificate_sds_secret_configs:
                name: client-certificate
                sds_config:
                  path: /etc/envoy/xds/sds-client-certificate.yaml
                  resource_api_version: V3
              tls_params:
                tls_minimum_protocol_version: TLSv1_2
              validation_context_sds_secret_config:
                name: client-validation
                sds_config:
                  path: /etc/envoy/xds/sds-client-validation.yaml
                  resource_api_version: V3
            sni: "service"
    type: STRICT_DNS
```

### PR DESCRIPTION
**Pull request by davinci26**
Commit Message:

Signed-off-by: Sotiris Nanopoulos <sonanopo@microsoft.com>

Fixes #15902

### Issue

The regression is coming from e7a0f7c

The root of the issue is that in order to generate the warning regarding using Admin tapping with UDS sockets we request the socket from the admin interface. This throws for the `config_validation` admin instance.

### Resolution

Add a private function to validation admin interface that initializes a dummy socket and this code passes.

Risk Level: Low (validation mode only)
Testing: Added config tests
Docs Changes: N/A
Release Notes: Fix an issue that causes TAP sockets to panic during config validation mode.

### CODE REVIEW

**Comment by davinci26** [comment]
@snowp do you want to do the review since you were involved in the issue? Otherwise I can bring in someone else

**Comment by davinci26** [comment]
cc @mattklein123 who added the warning on Tap

File: docs/root/version_history/current.rst [hunk 1]
```diff
@@ -77,7 +77,7 @@ Bug Fixes
 ---------
 *Changes expected to improve the state of the world and are unlikely to have negative effects*
 
-* active http health checks: properly handles HTTP/2 GOAWAY frames from the upstream. Previously a GOAWAY frame due to a graceful listener drain could cause improper failed health checks due to streams being refused by the upstream on a connection that is going away. To revert to old GOAWAY handling behavior, set the runtime feature `envoy.reloadable_features.health_check.graceful_goaway_handling` to false.
+* active http health checks: properly handles HTTP/2 GOAWAY frames from the upstream. Previously a GON/AAWAY frame due to a graceful listener drain could cause improper failed health checks due to streams being refused by the upstream on a connection that is going away. To revert to old GOAWAY handling behavior, set the runtime feature `envoy.reloadable_features.health_check.graceful_goaway_handling` to false.
```
Comment (snowp) [inline docs/root/version_history/current.rst:80]: revert?

File: source/server/config_validation/admin.h [hunk 2]
```diff
@@ -30,8 +32,15 @@ class ValidationAdmin : public Admin {
...
 
+  // We want to implement the socket interface without implementing the http listener function.
+  // This is useful for TAP because it wants to emit warnings when the address type is UDS
+  void createHttpListenerSocket(Network::Address::InstanceConstSharedPtr address) {
+    socket_ = std::make_shared<Network::TcpListenSocket>(nullptr, address, nullptr);
+  }
+
 private:
   ConfigTrackerImpl config_tracker_;
+  Network::SocketSharedPtr socket_;
```
Comment (snowp) [inline source/server/config_validation/admin.h:43]: Can we have the `ValidationAdmin` take the config as part of its ctor and initialize this there instead?

File: test/mocks/server/admin.cc [hunk 3]
```diff
@@ -13,6 +14,8 @@ MockAdmin::MockAdmin() {
   ON_CALL(*this, getConfigTracker()).WillByDefault(ReturnRef(config_tracker_));
   ON_CALL(*this, concurrency()).WillByDefault(Return(1));
   ON_CALL(*this, socket()).WillByDefault(ReturnRef(socket_));
+  ON_CALL(*this, addHandler(_, _, _, _, _)).WillByDefault(Return(true));
+  ON_CALL(*this, removeHandler(_)).WillByDefault(Return(true));
```
Comment (snowp) [inline test/mocks/server/admin.cc:18]: Are these changes necessary?

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks, just a few questions

File: docs/root/version_history/current.rst (hunk 1 above)
Comment (davinci26) [reply docs/root/version_history/current.rst:80]: yeah will revert

File: test/mocks/server/admin.cc (hunk 3 above)
Comment (davinci26) [reply test/mocks/server/admin.cc:18]: They are if you want to add a regression test, otherwise the `example_config_test` is failing because this returns false and the assert in the TAP code hits

File: source/server/config_validation/admin.h (hunk 2 above)
Comment (davinci26) [reply source/server/config_validation/admin.h:43]: No strong opinion, here. I will create a constructor and add it there. Admin was default constructed and I wanted to keep it that way.

File: test/mocks/server/admin.cc (hunk 3 above)
Comment (snowp) [reply test/mocks/server/admin.cc:18]: Okay no problem, the connection just wasn't immediately clear

Comment (davinci26) [reply test/mocks/server/admin.cc:18]: I know, the code for the `config_validation` is a bit tricky to read and it is not obvious that it depends on these mocks

File: source/server/config_validation/admin.h (hunk 2 above)
Comment (davinci26) [reply source/server/config_validation/admin.h]: Fixed in 2cf43a55597b14c77e62892549b89b4132c6dca9

File: docs/root/version_history/current.rst (hunk 1 above)
Comment (davinci26) [reply docs/root/version_history/current.rst]: fixed in 2cf43a55597b14c77e62892549b89b4132c6dca9

**Comment by wrowe** [comment]
I'd like to know why our repokitteh can't do post-submit tasks. (Cat-owner hint: cats don't do tasks :)

**Comment by davinci26** [comment]
Attempting to pick up main again. Honestly I do not see how this change might have broken the example verification.

**Comment by davinci26** [comment]
@snowp it took me some time fighting the CI but I expect it to pass now

File: docs/root/version_history/current.rst [hunk 4]
```diff
@@ -13,6 +13,34 @@ Bug Fixes
 ---------
 *Changes expected to improve the state of the world and are unlikely to have negative effects*
 
+* active http health checks: properly handles HTTP/2 GOAWAY frames from the upstream. Previously a GOAWAY frame due to a graceful listener drain could cause improper failed health checks due to streams being refused by the upstream on a connection that is going away. To revert to old GOAWAY handling behavior, set the runtime feature `envoy.reloadable_features.health_check.graceful_goaway_handling` to false.
+* adaptive concurrency: fixed a bug where concurrent requests on different worker threads could update minRTT back-to-back.
+* buffer: tighten network connection read and write buffer high watermarks in preparation to more careful enforcement of read limits. Buffer high-watermark is now set to the exact configured value; previously it was set to value + 1.
```
Comment (snowp) [inline docs/root/version_history/current.rst:18]: Seems like bad merge, can you sort this out?

File: source/server/config_validation/admin.h [hunk 5]
```diff
@@ -16,6 +18,11 @@ namespace Server {
  */
 class ValidationAdmin : public Admin {
 public:
+  // We want to implement the socket interface without implementing the http listener function.
+  // This is useful for TAP because it wants to emit warnings when the address type is UDS
+  ValidationAdmin(Network::Address::InstanceConstSharedPtr address)
```
Comment (snowp) [inline source/server/config_validation/admin.h:23]: `explicit`

File: source/server/config_validation/admin.h [hunk 6]
```diff
@@ -16,6 +18,11 @@ namespace Server {
...
+      : socket_(address ? std::make_shared<Network::TcpListenSocket>(nullptr, address, nullptr)
```
Comment (snowp) [inline source/server/config_validation/admin.h:24]: `std::move(address)`

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks this looks code overall, just some small comments

**Review by snowp — DISMISSED** [review]
LGTM, thanks!

Unfortunately I think you need another main merge

**Review by snowp — APPROVED** [review]
